components/pd_client/src/client_v2.rs

```diff
@@ -0,0 +1,1339 @@
+                f();
+            }
+        }
+        Ok(())
+    }
+
+    #[inline]
+    fn stub(&self) -> &PdClientStub {
+        &self.cache.stub
+    }
+
+    #[inline]
+    fn channel(&self) -> &Channel {
+        &self.cache.channel
+    }
+
+    #[inline]
+    fn cache_version(&self) -> u64 {
+        self.cache_version
+    }
+
+    #[inline]
+    fn call_option(&self) -> CallOption {
+        self.cache.target_info.call_option()
+    }
+
+    #[inline]
+    fn leader(&self) -> Member {
+        self.cache.members.get_leader().clone()
+    }
+}
+
+pub struct CachedDuplexResponse<T> {
+    latest: DuplexResponseHolder<T>,
+    cache: Option<ClientDuplexReceiver<T>>,
+}
+
+impl<T: Debug> Stream for CachedDuplexResponse<T> {
+    type Item = Result<T>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            if let Some(ref mut receiver) = self.cache {
+                match Pin::new(receiver).poll_next(cx) {
+                    Poll::Ready(Some(Ok(item))) => return Poll::Ready(Some(Ok(item))),
+                    Poll::Pending => return Poll::Pending,
+                    // If it's None or there's error, we need to update receiver.
+                    _ => {}
+                }
+            }
+
+            let latest = self.latest.lock().unwrap().take();
+            self.cache = latest;
+            if self.cache.is_none() {
+                return Poll::Pending;
+            }
+        }
+    }
+}
+
+type DuplexResponseHolder<T> = Arc<StdMutex<Option<ClientDuplexReceiver<T>>>>;
+type CallbackHolder = Arc<StdMutex<Option<Box<dyn Fn() + Sync + Send + 'static>>>>;
+
+#[derive(Clone)]
+pub struct RpcClient {
+    cluster_id: u64,
+    raw_client: CachedRawClient,
+    feature_gate: FeatureGate,
+    reconnect_tx: mpsc::UnboundedSender<u64>,
+    on_reconnect: CallbackHolder,
+}
+
+impl RpcClient {
+    pub fn new(
+        cfg: &Config,
+        shared_env: Option<Arc<Environment>>,
+        security_mgr: Arc<SecurityManager>,
+    ) -> Result<RpcClient> {
+        block_on(Self::new_async(cfg, shared_env, security_mgr))
+    }
+
+    pub async fn new_async(
+        cfg: &Config,
+        shared_env: Option<Arc<Environment>>,
+        security_mgr: Arc<SecurityManager>,
+    ) -> Result<RpcClient> {
+        let env = shared_env.unwrap_or_else(|| {
+            Arc::new(
+                EnvBuilder::new()
+                    .cq_count(CQ_COUNT)
+                    .name_prefix(thd_name!(CLIENT_PREFIX))
+                    .build(),
+            )
+        });
+
+        // -1 means the max.
+        let retries = match cfg.retry_max_count {
+            -1 => std::isize::MAX,
+            v => v.saturating_add(1),
+        };
+        let pd_connector = PdConnector::new(env.clone(), security_mgr.clone());
+        for i in 0..retries {
+            match pd_connector.validate_endpoints(cfg, false).await {
```

Comment:
Initialize connection should also be put into singleton connection coroutine.

Reply:
I still don't understand why, the code isn't any fewer with it stuffed into the coroutine.

Reply:
The goal is not to reduce lines of code, but to make the lifetime of a connection clear. Now it's split into different parts.

Reply:
Is there a way to spawn coroutine when the client doesn't exist yet? And are requests simply rejected if the client is not created yet? If so the behavior seems inferior.

Reply:
> Is there a way to spawn coroutine when the client doesn't exist yet

I don't see how coroutine and client are related. The shared part should be RawClient.

The request should not be rejected. It should call `wait_for_ready` and if it's timeout, error is returned. If caller expects the client should be ready before serving, then it can call `wait_for_ready` directly.

Reply:
`reconnect_loop` is spawned into grpc pool, which is accessed via a grpc client. The question is how to access that pool without a client in the first place.

Reply:
I see. Technically this can be solved by creating a client to arbitrary address. Though make `Env` be able to spawn a future is more natural solution.

Reply:
> It should call wait_for_ready and if it's timeout

How to wait for it exactly (without adding new synchronization primitive which IMO makes the code more complex with little benefit)? My early draft before uses the mutex to synchronize (i.e. lock the mutex and move it into initialization coroutine, so that consumer can wait on that mutex), which is too obscure so I went back with the blocking approach.

Reply:
I added a lame channel to grpcio, it should help merging first step with the rest. https://github.com/tikv/grpc-rs/pull/597